pkg/schedule/coordinator.go

```diff
@@ -152,72 +149,226 @@ func (c *Coordinator) IsPendingRegion(region uint64) bool {
+					<-c.patrolRegionContext.regionChan
+				}
+				log.Debug("skip patrol regions due to scheduling is halted")
+				continue
+			}
+
+			// wait for the regionChan to be drained
+			if len(c.patrolRegionContext.regionChan) > 0 {
+				continue
+			}
+
+			// Check priority regions first.
+			c.checkPriorityRegions()
+			// Check suspect regions first.
+			c.checkSuspectRegions()
+			// Check regions in the waiting list
+			c.checkWaitingRegions()
+
+			key, regions = c.checkRegions(key)
+			if len(regions) == 0 {
+				continue
+			}
+			// Updates the label level isolation statistics.
+			c.cluster.UpdateRegionsLabelLevelStats(regions)
+			// Update metrics and scan limit if a full scan is done.
+			if len(key) == 0 {
+				c.patrolRegionContext.roundUpdateMetrics()
+				c.patrolRegionContext.setScanLimit(calculateScanLimit(c.cluster))
+			}
+			failpoint.Inject("break-patrol", func() {
+				time.Sleep(100 * time.Millisecond) // ensure the regions are handled by the workers
+				failpoint.Return()
+			})
 		case <-c.ctx.Done():
 			patrolCheckRegionsGauge.Set(0)
-			c.setPatrolRegionsDuration(0)
+			c.patrolRegionContext.setPatrolRegionsDuration(0)
 			log.Info("patrol regions has been stopped")
 			return
 		}
-		if c.cluster.IsSchedulingHalted() {
-			continue
-		}
+	}
+}
 
-		// Check priority regions first.
-		c.checkPriorityRegions()
-		// Check suspect regions first.
-		c.checkSuspectRegions()
-		// Check regions in the waiting list
-		c.checkWaitingRegions()
+func (c *Coordinator) updateTickerIfNeeded(ticker *time.Ticker) {
+	// Note: we reset the ticker here to support updating configuration dynamically.
+	newInterval := c.cluster.GetCheckerConfig().GetPatrolRegionInterval()
+	if c.patrolRegionContext.getInterval() != newInterval {
+		c.patrolRegionContext.setInterval(newInterval)
+		ticker.Reset(newInterval)
+		log.Debug("coordinator starts patrol regions with new interval", zap.Duration("interval", newInterval))
+	}
+}
 
-		key, regions = c.checkRegions(key)
-		if len(regions) == 0 {
-			continue
-		}
-		// Updates the label level isolation statistics.
-		c.cluster.UpdateRegionsLabelLevelStats(regions)
-		if len(key) == 0 {
-			dur := time.Since(start)
-			patrolCheckRegionsGauge.Set(dur.Seconds())
-			c.setPatrolRegionsDuration(dur)
-			start = time.Now()
-		}
-		failpoint.Inject("break-patrol", func() {
-			failpoint.Break()
-		})
+func (c *Coordinator) updatePatrolWorkersIfNeeded() {
+	newWorkersCount := c.cluster.GetCheckerConfig().GetPatrolRegionWorkerCount()
+	if c.patrolRegionContext.getWorkerCount() != newWorkersCount {
+		oldWorkersCount := c.patrolRegionContext.getWorkerCount()
+		c.patrolRegionContext.setWorkerCount(newWorkersCount)
+		// Stop the old workers and start the new workers.
+		c.patrolRegionContext.workersCancel()
+		c.patrolRegionContext.wg.Wait()
+		c.patrolRegionContext.workersCtx, c.patrolRegionContext.workersCancel = context.WithCancel(c.ctx)
+		c.patrolRegionContext.startPatrolRegionWorkers(c)
```

Comment:
Only one thread to scan and multiple threads to process?

Reply:
yes